server/src/main/java/org/opensearch/index/codec/composite/Composite99DocValuesReader.java

```diff
@@ -29,11 +60,123 @@
+            readState.segmentInfo.name,
+            readState.segmentSuffix,
+            Composite99DocValuesFormat.META_EXTENSION
+        );
+
+        String dataFileName = IndexFileNames.segmentFileName(
+            readState.segmentInfo.name,
+            readState.segmentSuffix,
+            Composite99DocValuesFormat.DATA_EXTENSION
+        );
+
+        boolean success = false;
+        try {
+
+            // initialize meta input
+            dataIn = readState.directory.openInput(dataFileName, readState.context);
+            CodecUtil.checkIndexHeader(
+                dataIn,
+                Composite99DocValuesFormat.DATA_CODEC_NAME,
+                Composite99DocValuesFormat.VERSION_START,
+                Composite99DocValuesFormat.VERSION_CURRENT,
+                readState.segmentInfo.getId(),
+                readState.segmentSuffix
+            );
+
+            // initialize data input
+            metaIn = readState.directory.openChecksumInput(metaFileName, readState.context);
+            Throwable priorE = null;
+            try {
+                CodecUtil.checkIndexHeader(
+                    metaIn,
+                    Composite99DocValuesFormat.META_CODEC_NAME,
+                    Composite99DocValuesFormat.VERSION_START,
+                    Composite99DocValuesFormat.VERSION_CURRENT,
+                    readState.segmentInfo.getId(),
+                    readState.segmentSuffix
+                );
+
+                while (true) {
+
+                    // validate magic marker
+                    long magicMarker = metaIn.readLong();
+                    if (magicMarker == -1) {
+                        logger.info("EOF reached for composite index metadata");
+                        break;
+                    } else if (magicMarker < 0) {
+                        throw new CorruptIndexException("Unknown token encountered: " + magicMarker, metaIn);
+                    } else if (MAGIC_MARKER != magicMarker) {
+                        logger.error("Invalid composite field magic marker");
+                        throw new IOException("Invalid composite field magic marker");
+                    }
+
+                    // construct composite index metadata
+                    CompositeIndexMetadata compositeIndexMetadata = new CompositeIndexMetadata(metaIn);
+                    String compositeFieldName = compositeIndexMetadata.getCompositeFieldName();
+                    compositeFieldInfos.add(
+                        new CompositeIndexFieldInfo(compositeFieldName, compositeIndexMetadata.getCompositeFieldType())
+                    );
+
+                    switch (compositeIndexMetadata.getCompositeFieldType()) {
+                        case STAR_TREE:
+                            StarTreeMetadata starTreeMetadata = compositeIndexMetadata.getStarTreeMetadata();
+                            Tree starTree = new StarTree(dataIn, starTreeMetadata);
```

Comment:
We can call it 'compositeIndexIn' and just return the slice of the index for this field. 

Instead of initializing StarTree. And we can always read the star tree root etc during getField() method